In [30]:
import pandas as pd

In [31]:
df = pd.read_csv('../data/pokemon_data.csv')
df.head(10)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,1,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,2,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,3,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,3,VenusaurMega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
4,4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
5,5,Charmeleon,Fire,NaN,58,64,58,80,65,80,1,False
6,6,Charizard,Fire,Flying,78,84,78,109,85,100,1,False
7,6,CharizardMega Charizard X,Fire,Dragon,78,130,111,130,85,100,1,False
8,6,CharizardMega Charizard Y,Fire,Flying,78,104,78,159,115,100,1,False
9,7,Squirtle,Water,NaN,44,48,65,50,64,43,1,False


In [32]:
df[df['HP'] > df['Attack']].head(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
13,10,Caterpie,Bug,NaN,45,30,35,20,20,45,1,False
14,11,Metapod,Bug,NaN,50,20,55,25,25,30,1,False
15,12,Butterfree,Bug,Flying,60,45,50,90,80,70,1,False
16,13,Weedle,Bug,Poison,40,35,30,20,20,50,1,False
17,14,Kakuna,Bug,Poison,45,25,50,25,25,35,1,False


In [33]:
## filter only with one type( no type 2)

df[df['Type 2'].isnull()].head(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
4,4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
5,5,Charmeleon,Fire,NaN,58,64,58,80,65,80,1,False
9,7,Squirtle,Water,NaN,44,48,65,50,64,43,1,False
10,8,Wartortle,Water,NaN,59,63,80,65,80,58,1,False
11,9,Blastoise,Water,NaN,79,83,100,85,105,78,1,False


# with and without loc

Your two lines

## A) Boolean filtering (most common)

df[df['Type 1'] == 'Grass'].head(5)

## B) .loc version

df.loc[df['Type 1'] == 'Grass'].head(5)

more explicit pandas method

label-based indexer

allows row + column selection together



## Key answer first

    In this case, both return the same result

    BUT the way they work internally is different.

| Feature                   | `df[...]` | `df.loc[...]`  |
| ------------------------- | --------- | -------------- |
| readability               | simpler   | clearer intent |
| column selection          | limited   | powerful       |
| row + column filtering    | no        | yes            |
| recommended in interviews | yes       | YES (better)   |


# Important insight
This works:
```python
df[df['Type 1'] == 'Grass']
```
But this is more powerful:
```python
df.loc[df['Type 1'] == 'Grass', ['Name', 'HP']]
```
## .loc = full control

Why .loc is preferred in serious code

Because it avoids ambiguity:

df[...] → could be row or column confusion

df.loc → explicitly row + column selector

======================================================


## One important subtle difference

If you later do assignment:
```python
df.loc[df['Type 1'] == 'Grass', 'HP'] = 0
```
It sets the HP of all Grass-type Pokémon to 0

✔ modifies safely

But:
```python
df[df['Type 1'] == 'Grass']['HP'] = 0
```
may fail (chained indexing problem)

# Final rule

.loc is safer and more powerful

df[...] is shorter but limited

===========================================================
# To change multiple columns in pandas, you just pass a list of column names instead of a single column.

## Basic idea
```python
df.loc[rows, [col1, col2, col3]] = value
```
Example 1: change multiple columns to same value

```python
df.loc[df['Type 1'] == 'Grass', ['HP', 'Attack']] = 0
```
sets HP and Attack to 0 for Grass Pokémon

In [34]:
df.loc[df['Type 1'] ==  'Grass'].head(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,1,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,2,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,3,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,3,VenusaurMega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
48,43,Oddish,Grass,Poison,45,50,55,75,65,30,1,False


In [35]:
df[df['Type 1'] == 'Grass'].head(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,1,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,2,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,3,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,3,VenusaurMega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
48,43,Oddish,Grass,Poison,45,50,55,75,65,30,1,False


In [36]:
df.loc[:,'Type 1']

0        Grass
1        Grass
2        Grass
3        Grass
4         Fire
        ...   
795       Rock
796       Rock
797    Psychic
798    Psychic
799       Fire
Name: Type 1, Length: 800, dtype: str

In [37]:
df.loc[df['Type 1'] == 'Grass', 'HP']

0       45
1       60
2       80
3       80
48      45
      ... 
718     56
719     61
720     88
740     66
741    123
Name: HP, Length: 70, dtype: int64

In [38]:
df.loc[(df['Type 1'] == 'Grass') & (df['HP'] > 100), ['Name', 'HP']]

,Name,HP
652,Amoonguss,114
741,Gogoat,123


In [39]:
df.loc[df['Type 1'] == 'Fire'].head(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
4,4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
5,5,Charmeleon,Fire,NaN,58,64,58,80,65,80,1,False
6,6,Charizard,Fire,Flying,78,84,78,109,85,100,1,False
7,6,CharizardMega Charizard X,Fire,Dragon,78,130,111,130,85,100,1,False
8,6,CharizardMega Charizard Y,Fire,Flying,78,104,78,159,115,100,1,False


In [40]:
df.loc[df['HP'] > 100, ['Name', 'HP'] ].head(10)

,Name,HP
44,Jigglypuff,115
45,Wigglytuff,140
96,Muk,105
120,Rhydon,105
121,Chansey,250
123,Kangaskhan,105
124,KangaskhanMega Kangaskhan,105
142,Lapras,130
145,Vaporeon,130
155,Snorlax,160


Here are some pandas filtering challenges using a Pokémon dataset (assuming columns like `Name`, `Type 1`, `Type 2`, `HP`, `Attack`, `Defense`, `Speed`, `Legendary`, `Generation`).

Try writing the `.loc[]` filters yourself before checking the answers.

### Easy

1. Get all Grass Pokémon.
2. Get all Pokémon with HP greater than 100.
3. Get all Legendary Pokémon.
4. Get all Pokémon from Generation 1.
5. Get only the `Name` and `Attack` columns for Fire Pokémon.

```python
# Example format
df.loc[df['Type 1'] == 'Grass']
```



In [41]:
df.loc[(df['Type 1'] == 'Grass') | (df['Type 2'] == 'Grass')].sample(10).sort_values('Name' , ascending=True)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
167,153,Bayleef,Grass,NaN,60,62,80,63,80,60,2,False
271,251,Celebi,Psychic,Grass,100,100,100,100,100,100,2,False
111,103,Exeggutor,Grass,Psychic,95,95,85,125,65,55,1,False
651,590,Foongus,Grass,Poison,69,55,45,55,55,15,5,False
521,470,Leafeon,Grass,NaN,65,110,130,60,65,95,4,False
783,710,PumpkabooSuper Size,Ghost,Grass,59,66,70,44,55,41,6,False
719,651,Quilladin,Grass,NaN,61,78,95,56,58,57,6,False
554,495,Snivy,Grass,NaN,45,45,55,45,55,63,5,False
206,191,Sunkern,Grass,NaN,30,30,30,30,30,30,2,False
76,70,Weepinbell,Grass,Poison,65,90,50,85,45,55,1,False


In [42]:
df.loc[df['HP'] > 100].sample(4)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
712,646,KyuremWhite Kyurem,Dragon,Ice,125,120,90,170,100,95,5,True
96,89,Muk,Poison,NaN,105,105,75,65,100,50,1,False
711,646,KyuremBlack Kyurem,Dragon,Ice,125,170,100,120,90,95,5,True
589,530,Excadrill,Ground,Steel,110,135,60,50,65,88,5,False


In [43]:
df.loc[~df['Legendary']].sample(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
317,293,Whismur,Normal,NaN,64,51,23,51,23,28,3,False
471,424,Ambipom,Normal,NaN,75,100,66,60,66,115,4,False
487,439,Mime Jr.,Psychic,Fairy,20,25,45,70,90,60,4,False
678,617,Accelgor,Bug,NaN,80,70,40,100,60,145,5,False
159,147,Dratini,Dragon,NaN,41,64,45,50,50,50,1,False


In [44]:
df.loc[df['Legendary']].sample(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
545,487,GiratinaOrigin Forme,Ghost,Dragon,150,120,100,120,100,90,4,True
537,480,Uxie,Psychic,NaN,75,75,130,75,130,95,4,True
540,483,Dialga,Steel,Dragon,100,120,120,150,100,90,4,True
544,487,GiratinaAltered Forme,Ghost,Dragon,150,100,120,100,120,90,4,True
417,380,Latias,Dragon,Psychic,80,80,90,110,130,110,3,True


In [45]:
df.loc[df['Generation'] == 1].sample(4)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
20,16,Pidgey,Normal,Flying,40,45,40,35,35,56,1,False
128,119,Seaking,Water,NaN,80,92,65,65,80,68,1,False
9,7,Squirtle,Water,NaN,44,48,65,50,64,43,1,False
120,112,Rhydon,Ground,Rock,105,130,120,45,45,40,1,False


In [46]:
df.loc[(df['Type 1'] == 'Fire') | (df['Type 2'] == 'Fire'),['Name', 'Attack']]

,Name,Attack
4,Charmander,52
5,Charmeleon,64
6,Charizard,84
7,CharizardMega Charizard X,130
8,CharizardMega Charizard Y,104
...,...,...
730,Fletchinder,73
731,Talonflame,81
735,Litleo,50
736,Pyroar,68


# Medium
### Get all Water Pokémon with HP above 80.
### Get all Pokémon whose Attack is greater than Defense.
### Get all Pokémon that are either Fire or Electric.
### Get all non-Legendary Pokémon.
### Get Pokémon with Speed between 80 and 120.

In [47]:
df.loc[(df['HP'] > 80) & ((df['Type 1'] == 'Water') | (df['Type 2'] == 'Water'))].sample(10).sort_values('Name', ascending=True)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
199,184,Azumarill,Water,Fairy,100,50,80,60,80,50,2,False
440,395,Empoleon,Water,Steel,84,86,88,111,101,60,4,False
713,647,KeldeoOrdinary Forme,Water,Fighting,91,72,90,129,90,108,5,False
421,382,Kyogre,Water,NaN,100,100,90,150,140,90,3,True
142,131,Lapras,Water,Ice,130,85,80,85,95,60,1,False
399,364,Sealeo,Ice,Water,90,60,70,75,70,45,3,False
87,80,SlowbroMega Slowbro,Water,Psychic,95,75,180,130,80,30,1,False
85,79,Slowpoke,Water,Psychic,90,65,65,40,40,15,1,False
145,134,Vaporeon,Water,NaN,130,65,60,110,95,65,1,False
372,340,Whiscash,Water,Ground,110,78,73,76,71,60,3,False


In [48]:
df.loc[df['Attack'] > df['Defense']].sample(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
306,282,GardevoirMega Gardevoir,Psychic,Fairy,68,85,65,165,135,100,3,False
448,403,Shinx,Electric,NaN,45,65,34,40,34,45,4,False
268,248,TyranitarMega Tyranitar,Rock,Dark,100,164,150,95,120,71,2,False
158,146,Moltres,Fire,Flying,90,100,90,125,85,90,1,True
213,198,Murkrow,Dark,Flying,60,85,42,85,42,91,2,False


In [49]:
df.loc[
    (df['Type 1'].isin(['Fire', 'Electric'])) |
    (df['Type 2'].isin(['Fire', 'Electric']))
].sample(10).sort_values('Name')

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
196,181,AmpharosMega Ampharos,Electric,Dragon,90,95,105,165,110,45,2,False
64,59,Arcanine,Fire,NaN,90,110,80,100,80,95,1,False
616,555,DarmanitanZen Mode,Fire,Psychic,105,30,105,140,105,55,5,False
723,655,Delphox,Fire,Psychic,75,69,72,114,100,104,6,False
735,667,Litleo,Fire,Normal,62,50,58,73,54,72,6,False
436,391,Monferno,Fire,Fighting,64,78,52,78,52,81,4,False
170,156,Quilava,Fire,NaN,58,64,58,80,65,80,2,False
31,26,Raichu,Electric,NaN,60,90,55,90,80,110,1,False
553,494,Victini,Psychic,Fire,100,100,100,100,100,100,5,True
42,37,Vulpix,Fire,NaN,38,41,40,50,65,65,1,False


In [50]:
df.loc[(df['Speed'] > 80) & (df['Speed'] < 120)]
# ==================OR========================
df.loc[df['Speed'].between(80, 120)].sample(10).reset_index()

,index,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,2,3,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
1,43,38,Ninetales,Fire,NaN,73,76,75,81,100,100,1,False
2,521,470,Leafeon,Grass,NaN,65,110,130,60,65,95,4,False
3,706,643,Reshiram,Dragon,Fire,100,120,100,150,120,90,5,True
4,673,612,Haxorus,Dragon,NaN,76,147,90,60,70,97,5,False
5,423,383,Groudon,Ground,NaN,100,150,140,100,90,90,3,True
6,442,397,Staravia,Normal,Flying,55,75,50,40,40,80,4,False
7,548,490,Manaphy,Water,NaN,100,100,100,100,100,100,4,False
8,607,547,Whimsicott,Grass,Fairy,60,67,85,77,75,116,5,False
9,798,720,HoopaHoopa Unbound,Psychic,Dark,80,160,60,170,130,80,6,True


# Hard
### Get Pokémon that are Grass AND Poison.
### Get Pokémon with HP > 100 and Attack > 100.
### Get Pokémon that are Legendary OR have Attack > 150.
### Get Pokémon that are not Water type.
### Get Pokémon whose name starts with "Char".

In [51]:
df.loc[
    df['Type 1'].isin(['Grass', 'Poison']) &
    df['Type 2'].isin(['Grass', 'Poison'])
]

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,1,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,2,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,3,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,3,VenusaurMega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
48,43,Oddish,Grass,Poison,45,50,55,75,65,30,1,False
49,44,Gloom,Grass,Poison,60,65,70,85,75,40,1,False
50,45,Vileplume,Grass,Poison,75,80,85,110,90,50,1,False
75,69,Bellsprout,Grass,Poison,50,75,35,70,30,40,1,False
76,70,Weepinbell,Grass,Poison,65,90,50,85,45,55,1,False
77,71,Victreebel,Grass,Poison,80,105,65,100,70,70,1,False


In [67]:
df.loc[
    (df['HP'] > 100) &
    (df['Attack'] > 100)
].sample(8)


,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
589,530,Excadrill,Ground,Steel,110,135,60,50,65,88,5,False
321,297,Hariyama,Fighting,NaN,144,120,60,40,60,50,3,False
164,150,MewtwoMega Mewtwo Y,Psychic,NaN,106,150,70,194,120,140,1,True
500,450,Hippowdon,Ground,NaN,108,112,118,68,72,47,4,False
711,646,KyuremBlack Kyurem,Dragon,Ice,125,170,100,120,90,95,5,True
710,646,Kyurem,Dragon,Ice,125,130,90,130,90,95,5,True
793,717,Yveltal,Dark,Flying,126,131,95,131,98,99,6,True
313,289,Slaking,Normal,NaN,150,160,100,95,65,100,3,False


# filtering from the sub element of a  value

In [68]:
df.loc[df['Name'].str.startswith('Char')]

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
4,4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
5,5,Charmeleon,Fire,NaN,58,64,58,80,65,80,1,False
6,6,Charizard,Fire,Flying,78,84,78,109,85,100,1,False
7,6,CharizardMega Charizard X,Fire,Dragon,78,130,111,130,85,100,1,False
8,6,CharizardMega Charizard Y,Fire,Flying,78,104,78,159,115,100,1,False


In [71]:
df.loc[df['Name'].str.contains('Mega')].sample(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
19,15,BeedrillMega Beedrill,Bug,Poison,65,150,40,15,80,145,1,False
8,6,CharizardMega Charizard Y,Fire,Flying,78,104,78,159,115,100,1,False
336,308,MedichamMega Medicham,Fighting,Psychic,60,100,85,80,85,100,3,False
248,229,HoundoomMega Houndoom,Dark,Fire,75,90,90,140,90,115,2,False
23,18,PidgeotMega Pidgeot,Normal,Flying,83,80,80,135,80,121,1,False


# SAFER filtering with .loc and additional parameters


| Goal                    | Fix          |
| ----------------------- | ------------ |
| Case-insensitive search | `case=False` |
| Avoid NaN errors        | `na=False`   |

    case=False → ignores uppercase/lowercase differences || matches: Mega, mega, MEGA, mEGa

    na=False → treats NaN as “not matching”

If you do:
```python
df['Name'].str.contains('Mega')
```
and there are NaN values in Name, pandas returns:

    NaN
    True
    False
    NaN

It does not automatically drop them.

In [75]:
df.loc[df['Name'].str.contains('Mega',case=False, na=False)].sample(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
354,323,CameruptMega Camerupt,Fire,Ground,70,120,100,145,105,20,3,False
141,130,GyaradosMega Gyarados,Water,Dark,95,155,109,70,130,81,1,False
333,306,AggronMega Aggron,Steel,NaN,70,140,230,60,80,50,3,False
275,254,SceptileMega Sceptile,Grass,Dragon,70,110,75,145,85,145,3,False
426,384,RayquazaMega Rayquaza,Dragon,Flying,105,180,100,180,100,115,3,True
